### PR TITLE
[workload] Remove Microsoft.Maui.Extensions pack

### PR DESCRIPTION
**Pull request by pjcollins**
Reduces the number of nupkgs and msi files we produce and distribute by
moving the PackageReferences declared in Microsoft.Maui.Extensions to
BundledVersions.targets

### CODE REVIEW

**Review by mattleibow — APPROVED** [review]

**Review by jonathanpeppers — APPROVED** [review]

File: src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets
Comment (pjcollins) [inline src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets:65]: @mattleibow and @jonathanpeppers I had a question around `$(UseMauiEssentials)=true`, as I am not too familiar with where/how that is used. It seems that projects that set this property to true will now get `Microsoft.Extensions` package references by default through `Microsoft.Maui.Dependencies`, but this may not be any different from how things are set up today? Are there cases where a project could be using `$(UseMauiEssentials)=true` without them using the MAUI workload or also setting `$(UseMaui)` to true?

Comment (jonathanpeppers) [reply src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets:65]: If you did `dotnet new android` and then `UseMauiEssentials=true`, it would be like using Xamarin.Essentials without Xamarin.Forms.

With this change, you'd also get `Microsoft.Extensions` packages, but it seems like that would be OK?

Comment (mattleibow) [reply src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets:65]: Not sure... There is a lot of extra things now I suppose such as the logging, config and more.

I suppose the linker will remove it?

What happens when we switch to library packs and no longer use this nuget?

**Review by mattleibow — CHANGES_REQUESTED** [review]
Changing this to a non mergable state just to get feedback from @rmarinho and @Redth

**Review by rmarinho — APPROVED** [review]

File: src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets
Comment (jonathanpeppers) [reply src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets:65]: > I suppose the linker will remove it?

It looks like Microsoft.Extensions takes up ~148kb uncompressed in a `dotnet new maui` app. It doesn't look like they are marked as trimmable yet.

But I think if the app doesn't use them at all, they would get dropped? `csc` would exclude them?

> What happens when we switch to library packs and no longer use this nuget?

I think this change actually helps this, as you are getting rid of a pack.

**Review by mattleibow — APPROVED** [review]

**Comment by pjcollins** [comment]
I tried to get some more info around possible app size implications for the Essentials case. I took the example of creating a dotnet new android project, setting UseMauiEssentials=true, and adding a really simple Barometer example to the template.

File sizes for a Release build are near identical:
| Before Merge | After Merging Dependencies and Extensions |
| --------------- | -------------- |
| aab - 17,975,539 bytes | aab - 17,975,595 bytes |
| apk - 19,013,606 bytes | apk - 19,013,606 bytes | 

A binlog also seems to show that these extra references are being removed by ILLink in this scenario, and they are not included in the `@(ResolvedFileToPublish)` output of that target:

image

File: src/Workload/Microsoft.Maui.Dependencies/Microsoft.Maui.Dependencies.csproj
Comment (jonathanpeppers) [inline src/Workload/Microsoft.Maui.Dependencies/Microsoft.Maui.Dependencies.csproj:19]: @pjcollins another idea is to not put these here and instead...

File: src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets
Comment (jonathanpeppers) [inline src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets:65]: Put them here behind a condition:
```xml
<ItemGroup Condition=" '$(UseMaui)' == 'true' or '$(UseMauiCore)' == 'true' ">
```
Then your .apk size example wouldn't even get these packages at all.

It looks like this file is already setup to replace version numbers, so that part looks like it would work.

**Review by mattleibow — APPROVED** [review]

**Review by mattleibow — CHANGES_REQUESTED** [review]
What Peppers said. 🤗

File: src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets
Comment (pjcollins) [reply src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets]: Good idea, it seems like this should be functionally the same as before without an extra nupkg/msi.  These packages didn't declare IsImplicitlyDefined or a specific version before, but it seems like they should? I hard-coded the latest stable versions for now as it seems like the version wouldn't change often, it can be updated to do a package version string replace in the future if needed.

File: src/Workload/Microsoft.Maui.Dependencies/Microsoft.Maui.Dependencies.csproj
Comment (pjcollins) [reply src/Workload/Microsoft.Maui.Dependencies/Microsoft.Maui.Dependencies.csproj]: Should the same be done for Dependencies at this point? It seems like we shouldn't need a library pack declaring any external package references when we could declare them in our own targets?

File: src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets
Comment (jonathanpeppers) [reply src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets]: Yes, I would put `IsImplicitlyDefined`. We could probably get rid of Microsoft.Maui.Dependencies, too, but maybe try that in a future PR.

We probably should use these properties instead of hardcoding those:

https://github.com/dotnet/maui/blob/64432c450080efa99084f040c933c9011b066989/eng/Versions.props#L7-L9

At some point in .NET 7, they might be updated by Maestro.

Comment (pjcollins) [reply src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets]: Updated, I think this should be good to go. Similar changes for the Dependencies pack can happen in the future if we want.

**Comment by pjcollins** [comment]
@mattleibow @jonathanpeppers I'm not sure if we want to hold this until .NET 7 at this point? I think it should be ready now though.

**Review by mattleibow — APPROVED** [review]

**Comment by pjcollins** [comment]
@mattleibow @rmarinho @jonathanpeppers is main open for .NET 7 related things like this? I think we should be good to merge this, unless there is a separate .NET 7 branch that this should go to instead?

**Comment by jonathanpeppers** [comment]
Should we put it on the `net7.0` branch?

**Comment by pjcollins** [comment]
No conflicts after changing the base to net7.0 if we want to merge it there. I assume net7.0 will merge back into main eventually?

**Pull request merged**
